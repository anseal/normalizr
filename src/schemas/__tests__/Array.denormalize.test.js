// eslint-env jest
import { denormalize, normalize, schema } from '../../'

describe('ArraySchema denormalization', () => {
	describe('Object', () => {
		test('denormalizes a single entity', () => {
			expect(true).toBe(true)
			const cats = new schema.Entity('cats')
			const entities = {
				cats: {
					1: { id: 1, name: 'Milo' },
					2: { id: 2, name: 'Jake' },
				},
			}
			expect(denormalize([1, 2], [cats], entities)).toMatchSnapshot()
		})

		test('returns the input value if is not an array', () => {
			const filling = new schema.Entity('fillings')
			const taco = new schema.Entity('tacos', { fillings: [filling] })
			const entities = {
				tacos: {
					123: {
						id: '123',
						fillings: null,
					},
				},
			}

			expect(denormalize('123', taco, entities)).toMatchSnapshot()
		})
	})

	describe('Class', () => {
		test('denormalizes a single entity', () => {
			const cats = new schema.Entity('cats')
			const entities = {
				cats: {
					1: { id: 1, name: 'Milo' },
					2: { id: 2, name: 'Jake' },
				},
			}
			const catList = new schema.Array(cats)
			expect(denormalize([1, 2], catList, entities)).toMatchSnapshot()
		})

		test('denormalizes multiple entities', () => {
			const catSchema = new schema.Entity('cats')
			const peopleSchema = new schema.Entity('person')
			const listSchema = new schema.Array(
				{
					cats: catSchema,
					dogs: {},
					people: peopleSchema,
				},
				(input, _parent, _key) => input.type || 'dogs'
			)

			const entities = {
				cats: {
					123: {
						id: '123',
						type: 'cats',
					},
					456: {
						id: '456',
						type: 'cats',
					},
				},
				person: {
					123: {
						id: '123',
						type: 'people',
					},
				},
			}

			const input = [
				{ id: '123', schema: 'cats' },
				{ id: '123', schema: 'people' },
				{ id: { id: '789' }, schema: 'dogs' },
				{ id: '456', schema: 'cats' },
			]

			expect(denormalize(input, listSchema, entities)).toMatchSnapshot()
		})

		test('returns the input value if is not an array', () => {
			const filling = new schema.Entity('fillings')
			const fillings = new schema.Array(filling)
			const taco = new schema.Entity('tacos', { fillings })
			const entities = {
				tacos: {
					123: {
						id: '123',
						fillings: {},
					},
				},
			}

			expect(denormalize('123', taco, entities)).toMatchSnapshot()
		})

		test('does not assume mapping of schema to attribute values when schemaAttribute is not set', () => {
			const cats = new schema.Entity('cats')
			const catRecord = new schema.Object({
				cat: cats,
			})
			const catList = new schema.Array(catRecord)
			const input = [
				{ cat: { id: 1 }, id: 5 },
				{ cat: { id: 2 }, id: 6 },
			]
			const output = normalize(input, catList)
			expect(output).toMatchSnapshot()
			expect(denormalize(output.result, catList, output.entities)).toEqual(input)
		})
	})
})
