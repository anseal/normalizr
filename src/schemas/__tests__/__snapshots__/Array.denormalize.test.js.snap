// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArraySchema denormalization Class denormalizes a single entity 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Milo",
  },
  Object {
    "id": 2,
    "name": "Jake",
  },
]
`;

exports[`ArraySchema denormalization Class denormalizes multiple entities 1`] = `
Array [
  Object {
    "id": "123",
    "type": "cats",
  },
  Object {
    "id": "123",
    "type": "people",
  },
  Object {
    "id": "789",
  },
  Object {
    "id": "456",
    "type": "cats",
  },
]
`;

exports[`ArraySchema denormalization Class does not assume mapping of schema to attribute values when schemaAttribute is not set 1`] = `
Object {
  "entities": Object {
    "cats": Object {
      "1": Object {
        "id": 1,
      },
      "2": Object {
        "id": 2,
      },
    },
  },
  "result": Array [
    Object {
      "cat": 1,
      "id": 5,
    },
    Object {
      "cat": 2,
      "id": 6,
    },
  ],
}
`;

exports[`ArraySchema denormalization Class returns the input value if is not an array 1`] = `
Object {
  "fillings": Object {},
  "id": "123",
}
`;

exports[`ArraySchema denormalization Object denormalizes a single entity 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Milo",
  },
  Object {
    "id": 2,
    "name": "Jake",
  },
]
`;

exports[`ArraySchema denormalization Object returns the input value if is not an array 1`] = `
Object {
  "fillings": null,
  "id": "123",
}
`;
